internal/cluster/cluster.go

```diff
@@ -481,14 +481,17 @@ func (m *Manager) SkipMonitoring(c *common.Cluster) bool {
 
 func (m *Manager) initMonitorQueryGenerator() {
 	if m.monitorQueryGenerator == nil {
-		noNeedToMonitorInStates := []string{
-			models.ClusterStatusInstalled,
-		}
+		prepareDB := func(db *gorm.DB) *gorm.DB {
+			noNeedToMonitorInStates := []string{
+				models.ClusterStatusInstalled,
+			}
 
-		dbWithCondition := common.LoadTableFromDB(m.db, common.HostsTable, "status <> ?", models.HostStatusDisabled)
-		dbWithCondition = common.LoadClusterTablesFromDB(dbWithCondition, common.HostsTable)
-		dbWithCondition = dbWithCondition.Where("status NOT IN (?)", noNeedToMonitorInStates)
-		m.monitorQueryGenerator = common.NewMonitorQueryGenerator(m.db, dbWithCondition, m.MonitorBatchSize)
+			dbWithCondition := common.LoadTableFromDB(db, common.HostsTable, "status <> ?", models.HostStatusDisabled)
+			dbWithCondition = common.LoadClusterTablesFromDB(dbWithCondition, common.HostsTable)
+			dbWithCondition = dbWithCondition.Where("status NOT IN (?)", noNeedToMonitorInStates)
+			return dbWithCondition
+		}
+		m.monitorQueryGenerator = common.NewMonitorQueryGenerator(m.db, prepareDB, m.MonitorBatchSize)
```

Comment:
As I wrote in the PR 
**Some chained methods of gorm.DB may not be reused.  For example
the value of db.Unscoped() cannot be used more than once.  Each such
invocation needs to call Unscoped() again.  The known chained
methods are: Unscoped, Where, Limit, Order, Clauses**

In this case we have a reused where clause.  As of now, it cannot be used this way in GORM v2.  So I had to change it.
Please see the issue in gorm https://github.com/go-gorm/gorm/issues/4847

Reply:
I guess I'm not sure how this avoids the issue.
In `prepareDB` we're doing this:

```
dbWithCondition = dbWithCondition.Where("status NOT IN (?)", noNeedToMonitorInStates)
return dbWithCondition
```

But then down [here](https://github.com/openshift/assisted-service/pull/2908/files/e9df23107a7219096ff7ca07f18a3491235375aa#diff-ae8489a6f9dce351a6ca906a090da834a0dba24cbb9a61221c05b6ac9b257935R116) you do:
```
err = t.prepareDB(t.db).Where("id in (?)", t.ids[t.offset:nextOffset]).Find(&clusters).Error
```
which chains another `Where` call onto the same db instance. Isn't this what you were trying to avoid?
I may not fully understand the issue though, which is why I asked.

Reply:
The where condition cannot be used more than once, so every time the function is called it is created again.
The previous version had a pre-created where condition that was reused many times.

Reply:
Ah I understand now.
It continues to build up the scope even after the query is executed.

This test run from the issue you created cleared up what was going on https://github.com/go-gorm/playground/runs/4178211401?check_suite_focus=true

Reply:
And this avoids the issue by applying the scope each time we need it rather than once at the start. Got it. Thanks @ori-amizur